File: backend/app/schemas/admin_partner_offer.py
"""Admin partner offer schemas — moderation queue (TICKET-305 / 305A)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.core.passport_constants import (
    DEFAULT_MAX_REDEMPTIONS_PER_PASSPORT,
    PARTNER_OFFER_TYPES,
    PartnerOfferStatus,
    PartnerOfferType,
)
from app.schemas.partner_offer_management import (
    PARTNER_OFFER_LIST_PAGE_SIZE_DEFAULT,
    PARTNER_OFFER_LIST_PAGE_SIZE_MAX,
)

__all__ = [
    "PARTNER_OFFER_LIST_PAGE_SIZE_DEFAULT",
    "PARTNER_OFFER_LIST_PAGE_SIZE_MAX",
    "PartnerOfferAdminCreateRequest",
    "PartnerOfferAdminUpdateRequest",
    "PartnerOfferAdminResponse",
    "PartnerOfferAdminListResponse",
    "VerifiedOrganizationOption",
    "VerifiedOrganizationListResponse",
    "PartnerOfferRejectRequest",
]


class PartnerOfferRejectRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str = Field(min_length=1, max_length=1000)


class PartnerOfferAdminCreateRequest(BaseModel):
    """Staff bootstrap — creates draft offers for verified organizations."""

    model_config = ConfigDict(extra="forbid")

    organization_id: UUID
    title: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)
    offer_type: PartnerOfferType
    valid_from: datetime | None = None
    valid_until: datetime | None = None
    redemption_limit: int = Field(default=DEFAULT_MAX_REDEMPTIONS_PER_PASSPORT, ge=1)
    max_redemptions_total: int | None = Field(default=None, ge=1)
    tier_code_required: str | None = Field(default=None, max_length=32)

    @field_validator("offer_type", mode="before")
    @classmethod
    def validate_offer_type(cls, value: object) -> object:
        if isinstance(value, str) and value not in PARTNER_OFFER_TYPES:
            raise ValueError("Type d'offre invalide.")
        return value

    @model_validator(mode="after")
    def validate_dates(self) -> PartnerOfferAdminCreateRequest:
        if self.valid_from and self.valid_until and self.valid_until <= self.valid_from:
            raise ValueError("La date de fin doit être postérieure à la date de début.")
        return self


class PartnerOfferAdminUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)
    offer_type: PartnerOfferType | None = None
    valid_from: datetime | None = None
    valid_until: datetime | None = None
    redemption_limit: int | None = Field(default=None, ge=1)
    max_redemptions_total: int | None = Field(default=None, ge=1)
    tier_code_required: str | None = Field(default=None, max_length=32)

    @model_validator(mode="after")
    def validate_dates(self) -> PartnerOfferAdminUpdateRequest:
        if self.valid_from and self.valid_until and self.valid_until <= self.valid_from:
            raise ValueError("La date de fin doit être postérieure à la date de début.")
        return self


class PartnerOfferOrganizationAdmin(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    name: str
    city: str
    verification_status: str
    visibility: str


class PartnerOfferAdminResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    title: str
    description: str | None
    offer_type: PartnerOfferType
    offer_status: PartnerOfferStatus
    is_active: bool
    tier_code_required: str | None
    max_redemptions_total: int | None
    redemption_limit: int
    valid_from: datetime | None
    valid_until: datetime | None
    redemptions_count: int
    created_by_user_id: UUID | None
    moderated_by_user_id: UUID | None
    moderated_at: datetime | None
    rejection_reason: str | None
    created_at: datetime
    updated_at: datetime
    organization: PartnerOfferOrganizationAdmin


class PartnerOfferAdminListResponse(BaseModel):
    items: list[PartnerOfferAdminResponse]
    total: int
    page: int
    page_size: int


class VerifiedOrganizationOption(BaseModel):
    id: UUID
    slug: str
    name: str
    city: str
    visibility: str


class VerifiedOrganizationListResponse(BaseModel):
    items: list[VerifiedOrganizationOption]

File: frontend/packages/types/src/admin_partner_offer.ts
import type { PartnerOfferType } from "./passport";
import type { PartnerOfferReadinessFields, PartnerOfferReadinessLevel } from "./partner-offer-readiness";
import type { OfferRedemptionStatus } from "./passport";
import type { PartnerOfferStatus } from "./partner_offer_management";

/** Alias admin — même enum produit que self-service (TICKET-305A). */
export type PartnerOfferAdminStatus = PartnerOfferStatus;

export interface PartnerOfferAdminOrganization {
  id: string;
  slug: string;
  name: string;
  city: string;
  verification_status: string;
  visibility: string;
}

export interface PartnerOfferAdmin {
  id: string;
  organization_id: string;
  title: string;
  description: string | null;
  value_label?: string | null;
  conditions?: string | null;
  offer_type: PartnerOfferType;
  offer_status: PartnerOfferAdminStatus;
  is_active: boolean;
  tier_code_required: string | null;
  max_redemptions_total: number | null;
  redemption_limit: number;
  valid_from: string | null;
  valid_until: string | null;
  redemptions_count: number;
  created_by_user_id: string | null;
  moderated_by_user_id: string | null;
  moderated_at: string | null;
  rejection_reason: string | null;
  created_at: string;
  updated_at: string;
  organization: PartnerOfferAdminOrganization;
  readiness: PartnerOfferReadinessFields;
}

export interface PartnerOfferAdminListResponse {
  items: PartnerOfferAdmin[];
  total: number;
  page: number;
  page_size: number;
}

export interface PartnerOfferAdminCreatePayload {
  organization_id: string;
  title: string;
  description?: string | null;
  offer_type: PartnerOfferType;
  valid_from?: string | null;
  valid_until?: string | null;
  redemption_limit?: number;
  max_redemptions_total?: number | null;
  tier_code_required?: string | null;
}

export interface PartnerOfferAdminUpdatePayload {
  title?: string;
  description?: string | null;
  offer_type?: PartnerOfferType;
  valid_from?: string | null;
  valid_until?: string | null;
  redemption_limit?: number;
  max_redemptions_total?: number | null;
  tier_code_required?: string | null;
}

export interface PartnerOfferRejectPayload {
  reason: string;
}

export interface VerifiedOrganizationOption {
  id: string;
  slug: string;
  name: string;
  city: string;
  visibility: string;
}

export interface VerifiedOrganizationListResponse {
  items: VerifiedOrganizationOption[];
}

export interface PartnerOfferAdminListParams {
  status?: PartnerOfferAdminStatus;
  offer_type?: PartnerOfferType;
  organization_id?: string;
  q?: string;
  readiness?: PartnerOfferReadinessLevel;
  page?: number;
  page_size?: number;
}

export interface PartnerOfferAdminSummaryResponse {
  city: string;
  generated_at: string;
  total: number;
  pending_review: number;
  published: number;
  draft: number;
  rejected: number;
  archived: number;
  contributor_partners: number;
  expired_or_inactive: number;
}

export interface PartnerOfferAdminSummaryParams {
  city?: string;
}

export type AdminOfferRedemptionChannel = "self" | "scan" | "unknown";

export interface AdminOfferRedemptionPassport {
  id: string;
  passport_number: string;
}

export interface AdminOfferRedemptionCitizen {
  id: string;
  display_name: string | null;
  email: string;
}

export interface PartnerOfferAdminRedemptionItem {
  id: string;
  passport: AdminOfferRedemptionPassport;
  citizen: AdminOfferRedemptionCitizen;
  channel: AdminOfferRedemptionChannel;
  status: OfferRedemptionStatus;
  redeemed_at: string | null;
}

export interface PartnerOfferAdminRedemptionListResponse {
  items: PartnerOfferAdminRedemptionItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface PartnerOfferAdminRedemptionListParams {
  page?: number;
  page_size?: number;
}

export type AdminPartnerOfferAction = "approve" | "reject" | "archive";

export interface AdminPartnerOfferActorSummary {
  id: string;
  email: string;
  display_name: string | null;
}

export interface AdminPartnerOfferActionItem {
  id: string;
  action: AdminPartnerOfferAction;
  previous_status: string | null;
  new_status: string | null;
  reason: string | null;
  actor_user: AdminPartnerOfferActorSummary;
  created_at: string;
}

export interface AdminPartnerOfferActionListResponse {
  items: AdminPartnerOfferActionItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface PartnerOfferAdminActionListParams {
  page?: number;
  page_size?: number;
}
